### PR TITLE
Capistrano3

### PR DESCRIPTION
**Pull request by pcrum**
Branch adds the ability for final release to maintain itself as a git repository versus utilizing git archive to deploy the final release.

Also added 'rbenv_install' capistrano plugin which includes the ability to install rbenv, bundle, and ruby.

### CODE REVIEW

File: config/deploy.rb [hunk 1]
```diff
@@ -144,12 +145,12 @@
  after :symlink_shared, :remove_resque
 
  # Restart the application
- desc 'Restart application'
- task :restart do
-  on roles(:app), in: :sequence, wait: 5 do
-      execute :touch, release_path.join('tmp/restart.txt')
-  end
- end
+ #desc 'Restart application'
```
Comment (mjgiarlo) [inline config/deploy.rb]: Feel free to delete these lines if we're not using them; we can always use git to pull them back later if need be.

File: db/schema.rb [hunk 2]
```diff
@@ -190,8 +190,8 @@
     t.string   "label"
     t.string   "lowerLabel"
     t.string   "uri"
-    t.datetime "created_at", default: '2014-05-12 17:29:21'
-    t.datetime "updated_at", default: '2014-05-12 17:29:21'
+    t.datetime "created_at", default: '2014-10-28 15:10:55'
```
Comment (mjgiarlo) [inline db/schema.rb]: Do we need changes to the schema to support capistrano3?

File: lib/capistrano/submodule_strategy.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class Capistrano::Git < Capistrano::SCM
+  module SubmoduleStrategy
+    include DefaultStrategy
+
+    # check for a .git directory
+  def test
+    test! " [ -d #{repo_path}/.git ] "
+  end
+ 
+  # same as in Capistrano::Git::DefaultStrategy
```
Comment (mjgiarlo) [inline lib/capistrano/submodule_strategy.rb]: If this method is the same as one pulled in via the include, do we need it in here?

File: lib/capistrano/submodule_strategy.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  def check
+    test! :git, :'ls-remote', repo_url
+  end
+ 
+  def clone
+    git :clone, '-b', fetch(:branch), '--recursive', repo_url, repo_path
+  end
+ 
+  # same as in Capistrano::Git::DefaultStrategy
```
Comment (mjgiarlo) [inline lib/capistrano/submodule_strategy.rb]: If this method is the same as one pulled in via the include, do we need it in here?

File: lib/capistrano/submodule_strategy.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  def update
+    git :remote, :update
+  end
+ 
+    def release
+      context.execute :rm, '-rf', release_path
```
Comment (mjgiarlo) [inline lib/capistrano/submodule_strategy.rb:14]: What's this deleting?

**Comment by mjgiarlo** [comment]
Do we need the empty .capignore file?

**Comment by mjgiarlo** [comment]
Hey @pcrum!  I left a handful of comments, and Travis is currently showing failing tests.  Once you get tests passing, can you squash your commits?  Thanks for the PR.

File: config/deploy.rb (hunk 1 above)
Comment (pcrum) [inline config/deploy.rb]: Not a problem, will do

File: db/schema.rb (hunk 2 above)
Comment (pcrum) [inline db/schema.rb]: No, no schema changes are needed. May have to review when this was added.

File: lib/capistrano/submodule_strategy.rb (hunk 3 above)
Comment (pcrum) [inline lib/capistrano/submodule_strategy.rb]: This module overrides the default git deploy strategy of capistrano 3 where the release path deployment is set using git archive where this overrides the method release method to remove the current release path and then clone the repo, branch, and any recursive submodules that may exist.

File: lib/capistrano/submodule_strategy.rb (hunk 4 above)
Comment (pcrum) [inline lib/capistrano/submodule_strategy.rb]: I think some of the methods are the same as the default but the important method override is the release method.

File: config/deploy.rb (hunk 1 above)
Comment (mjgiarlo) [inline config/deploy.rb]: Thx!  I love :bomb:ing code. :)

File: db/schema.rb (hunk 2 above)
Comment (mjgiarlo) [inline db/schema.rb]: K.

File: lib/capistrano/submodule_strategy.rb (hunk 3 above)
Comment (mjgiarlo) [inline lib/capistrano/submodule_strategy.rb]: OK. So we should only need to define methods that differ from the implementations already provided by the module, and can remove the others marked as "same as?"

**Comment by pcrum** [comment]
Hey @mjgiarlo , I believe I've corrected the formatting. Please take a look when you have a moment. Thanks

**Comment by mjgiarlo** [comment]
Looks great, @pcrum!  I believe the Travis error above is unrelated to this work, so I'm inclined to merge this.  Thoughts, @cam156 @hectorcorrea @awead @mtribone?

**Comment by hectorcorrea** [comment]
@mjgiarlo @pcrum I agree that the error does not seem to be related to the changes. I've restarted the build and will check again in 20 or so minutes.

**Comment by mjgiarlo** [comment]
:clap:

**Comment by awead** [comment]
Looks good to me, I'm pulling the trigger.

**Comment by mjgiarlo** [comment]
:+1: @pcrum

**Pull request merged**